**Linux** [Build 9644](http://brontes.lab.eng.brq.redhat.com/viewLog.html?buildId=94489&buildTypeId=WFPR) outcome was **FAILURE** using a merge of cc5977ac53dfbdd634b5f9c1fcd30f314f4c8271
Summary: Tests failed: 1 (1 new), passed: 3317, ignored: 194 Build time: 00:57:19
### Failed tests
```
org.jboss.as.test.integration.management.deploy.runtime.StatefulEJBRemoteHomeRuntimeNameTestCase.testStatefulLocalHome: java.rmi.RemoteException: Error; nested exception is: 
 java.io.IOException: Channel Channel ID fbcc4c7f (outbound) of Remoting connection 0128e61e to /0:0:0:0:0:0:0:1:8080 has been closed
java.rmi.RemoteException: Error; nested exception is: 
	java.io.IOException: Channel Channel ID fbcc4c7f (outbound) of Remoting connection 0128e61e to /0:0:0:0:0:0:0:1:8080 has been closed
	at org.jboss.ejb.client.EJBInvocationHandler.doInvoke(EJBInvocationHandler.java:236)
	at org.jboss.ejb.client.EJBInvocationHandler.doInvoke(EJBInvocationHandler.java:183)
	at org.jboss.ejb.client.EJBInvocationHandler.invoke(EJBInvocationHandler.java:146)
	at com.sun.proxy.$Proxy128.getEJBHome(Unknown Source)
	at org.jboss.as.test.integration.management.deploy.runtime.StatefulEJBRemoteHomeRuntimeNameTestCase.testStatefulLocalHome(StatefulEJBRemoteHomeRuntimeNameTestCase.java:145)
Caused by: java.io.IOException: Channel Channel ID fbcc4c7f (outbound) of Remoting connection 0128e61e to /0:0:0:0:0:0:0:1:8080 has been closed
	at org.jboss.ejb.client.remoting.ChannelAssociation$1.handleClose(ChannelAssociation.java:123)
	at org.jboss.ejb.client.remoting.ChannelAssociation$1.handleClose(ChannelAssociation.java:115)
	at org.jboss.remoting3.spi.SpiUtils.safeHandleClose(SpiUtils.java:54)
	at org.jboss.remoting3.spi.AbstractHandleableCloseable$CloseHandlerTask.run(AbstractHandleableCloseable.java:514)
	at org.jboss.remoting3.spi.AbstractHandleableCloseable.runCloseTask(AbstractHandleableCloseable.java:419)
	at org.jboss.remoting3.spi.AbstractHandleableCloseable.closeComplete(AbstractHandleableCloseable.java:290)
	at org.jboss.remoting3.remote.RemoteConnectionChannel.closeReadsAndWrites(RemoteConnectionChannel.java:274)
	at org.jboss.remoting3.remote.RemoteConnectionChannel.closeAction(RemoteConnectionChannel.java:534)
	at org.jboss.remoting3.spi.AbstractHandleableCloseable.closeAsync(AbstractHandleableCloseable.java:372)
	at org.jboss.remoting3.remote.RemoteConnectionHandler.closeAllChannels(RemoteConnectionHandler.java:437)
	at org.jboss.remoting3.remote.RemoteConnectionHandler.sendCloseRequest(RemoteConnectionHandler.java:236)
	at org.jboss.remoting3.remote.RemoteConnectionHandler.closeAction(RemoteConnectionHandler.java:413)
	at org.jboss.remoting3.spi.AbstractHandleableCloseable.closeAsync(AbstractHandleableCloseable.java:372)
	at org.jboss.remoting3.ConnectionImpl.closeAction(ConnectionImpl.java:54)
	at org.jboss.remoting3.spi.AbstractHandleableCloseable.close(AbstractHandleableCloseable.java:154)
	at org.jboss.ejb.client.remoting.ConnectionPool.safeClose(ConnectionPool.java:177)
	at org.jboss.ejb.client.remoting.ConnectionPool.release(ConnectionPool.java:104)
	at org.jboss.ejb.client.remoting.ConnectionPool$PooledConnection.close(ConnectionPool.java:198)
	at org.jboss.ejb.client.remoting.RemotingConnectionManager.safeClose(RemotingConnectionManager.java:65)
	at org.jboss.ejb.client.remoting.ConfigBasedEJBClientContextSelector$ContextCloseListener.contextClosed(ConfigBasedEJBClientContextSelector.java:220)
	at org.jboss.ejb.client.EJBClientContext.close(EJBClientContext.java:1305)
	at org.jboss.ejb.client.EJBClientContext.finalize(EJBClientContext.java:1316)
	at java.lang.System$2.invokeFinalize(System.java:1270)
	at java.lang.ref.Finalizer.runFinalizer(Finalizer.java:98)
	at java.lang.ref.Finalizer.access$100(Finalizer.java:34)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:210)
------- Stdout: -------
23:40:31,930 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-3) EJBCLIENT000017: Received server version 2 and marshalling strategies [river]
23:40:31,930 INFO  [org.jboss.ejb.client.remoting] (ejb-client-context-tasks-12-thread-7) EJBCLIENT000013: Successful version handshake completed for receiver context EJBReceiverContext{clientContext=org.jboss.ejb.client.EJBClientContext@598ef7, receiver=Remoting connection EJB receiver [connection=org.jboss.ejb.client.remoting.ConnectionPool$PooledConnection@445e47,channel=jboss.ejb,nodename=build25-el7]} on channel Channel ID fbcc4c7f (outbound) of Remoting connection 0128e61e to /0:0:0:0:0:0:0:1:8080
23:40:32,159 INFO  [org.jboss.ejb.client.remoting] (Remoting "config-based-ejb-client-endpoint" task-7) EJBCLIENT000016: Channel Channel ID fbcc4c7f (outbound) of Remoting connection 0128e61e to /0:0:0:0:0:0:0:1:8080 can no longer process messages
 [31m23:40:32,161 ERROR [org.jboss.as.ejb3.remote] (default task-25) WFLYEJB0150: Could not write method invocation failure for method public abstract javax.ejb.EJBHome javax.ejb.EJBObject.getEJBHome() throws java.rmi.RemoteException on bean named SimpleStatefulBean for appname nooma-nooma1-stateful-session-bean modulename ejb distinctname  due to: java.io.IOException: WFLYEJB0422: Could not open message outputstream for writing to Channel
	at org.jboss.as.ejb3.remote.protocol.versionone.MethodInvocationMessageHandler.writeMethodInvocationResponse(MethodInvocationMessageHandler.java:360)
	at org.jboss.as.ejb3.remote.protocol.versionone.MethodInvocationMessageHandler.access$600(MethodInvocationMessageHandler.java:67)
	at org.jboss.as.ejb3.remote.protocol.versionone.MethodInvocationMessageHandler$1.run(MethodInvocationMessageHandler.java:246)
	at org.jboss.as.ejb3.remote.protocol.versionone.MethodInvocationMessageHandler.processMessage(MethodInvocationMessageHandler.java:262)
	at org.jboss.as.ejb3.remote.protocol.versionone.VersionOneProtocolChannelReceiver.processMessage(VersionOneProtocolChannelReceiver.java:213)
	at org.jboss.as.ejb3.remote.protocol.versiontwo.VersionTwoProtocolChannelReceiver.processMessage(VersionTwoProtocolChannelReceiver.java:76)
	at org.jboss.as.ejb3.remote.protocol.versionone.VersionOneProtocolChannelReceiver.handleMessage(VersionOneProtocolChannelReceiver.java:159)
	at org.jboss.remoting3.remote.RemoteConnectionChannel$5.run(RemoteConnectionChannel.java:456)
	at org.jboss.remoting3.EndpointImpl$TrackingExecutor$1.run(EndpointImpl.java:717)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)


```

--- a/clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/ClassLoaderThreadFactory.java
+++ b/clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/ClassLoaderThreadFactory.java
@@ -1,0 +1,64 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.clustering.infinispan;
+
+import java.security.PrivilegedAction;
+import java.util.concurrent.ThreadFactory;
+
+import org.wildfly.security.manager.WildFlySecurityManager;
+
+/**
+ * {@link ThreadFactory} decorator that associates a specific class loader to created threads.
+ * @author Paul Ferraro
+ */
+public class ClassLoaderThreadFactory implements ThreadFactory {
+
+    private final ThreadFactory factory;
+    private final ClassLoader loader;
+
+    public ClassLoaderThreadFactory(ThreadFactory factory, ClassLoader loader) {
+        this.factory = factory;
+        this.loader = loader;
+    }
+
+    @Override
+    public Thread newThread(Runnable r) {
+        Runnable task = () -> {
+            try {
+                r.run();
+            } finally {
+                // Defensively reset the TCCL
+                this.setContextClassLoader(Thread.currentThread());
+            }
+        };
+        return this.setContextClassLoader(this.factory.newThread(task));
+    }
+
+    private Thread setContextClassLoader(Thread thread) {
+        PrivilegedAction<Thread> action = () -> {
+            thread.setContextClassLoader(this.loader);
+            return thread;
+        };
+        return WildFlySecurityManager.doUnchecked(action);
+    }
+}

--- a/clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ScheduledThreadPoolBuilder.java
+++ b/clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ScheduledThreadPoolBuilder.java
@@ -31,6 +31,7 @@ import org.infinispan.commons.executors.ThreadPoolExecutorFactory;
 import org.infinispan.configuration.global.ThreadPoolConfiguration;
 import org.infinispan.configuration.global.ThreadPoolConfigurationBuilder;
 import org.jboss.as.clustering.controller.ResourceServiceBuilder;
+import org.jboss.as.clustering.infinispan.ClassLoaderThreadFactory;
 import org.jboss.as.controller.OperationContext;
 import org.jboss.as.controller.OperationFailedException;
 import org.jboss.dmr.ModelNode;
@@ -60,7 +61,7 @@ public class ScheduledThreadPoolBuilder extends CacheContainerComponentBuilder<T
             @Override
             public ScheduledExecutorService createExecutor(ThreadFactory factory) {
                 // Use fixed size, based on maxThreads
-                ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(maxThreads, factory);
+                ScheduledThreadPoolExecutor executor = new ScheduledThreadPoolExecutor(maxThreads, new ClassLoaderThreadFactory(factory, ClassLoaderThreadFactory.class.getClassLoader()));
                 executor.setKeepAliveTime(keepAliveTime, TimeUnit.MILLISECONDS);
                 executor.setRemoveOnCancelPolicy(true);
                 executor.setExecuteExistingDelayedTasksAfterShutdownPolicy(false);
@@ -69,7 +70,7 @@ public class ScheduledThreadPoolBuilder extends CacheContainerComponentBuilder<T
 
             @Override
             public void validate() {
-
+                // Do nothing
             }
         };
         this.builder.threadPoolFactory(factory);

--- a/clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolBuilder.java
+++ b/clustering/infinispan/extension/src/main/java/org/jboss/as/clustering/infinispan/subsystem/ThreadPoolBuilder.java
@@ -22,11 +22,15 @@
 
 package org.jboss.as.clustering.infinispan.subsystem;
 
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ThreadFactory;
+
 import org.infinispan.commons.executors.BlockingThreadPoolExecutorFactory;
 import org.infinispan.commons.executors.ThreadPoolExecutorFactory;
 import org.infinispan.configuration.global.ThreadPoolConfiguration;
 import org.infinispan.configuration.global.ThreadPoolConfigurationBuilder;
 import org.jboss.as.clustering.controller.ResourceServiceBuilder;
+import org.jboss.as.clustering.infinispan.ClassLoaderThreadFactory;
 import org.jboss.as.controller.OperationContext;
 import org.jboss.as.controller.OperationFailedException;
 import org.jboss.dmr.ModelNode;
@@ -53,7 +57,12 @@ public class ThreadPoolBuilder extends CacheContainerComponentBuilder<ThreadPool
                 this.definition.getMinThreads().getDefinition().resolveModelAttribute(context, model).asInt(),
                 this.definition.getQueueLength().getDefinition().resolveModelAttribute(context, model).asInt(),
                 this.definition.getKeepAliveTime().getDefinition().resolveModelAttribute(context, model).asLong()
-        );
+        ) {
+            @Override
+            public ExecutorService createExecutor(ThreadFactory factory) {
+                return super.createExecutor(new ClassLoaderThreadFactory(factory, ClassLoaderThreadFactory.class.getClassLoader()));
+            }
+        };
         this.builder.threadPoolFactory(factory);
 
         return this;

--- a/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/ClassLoaderThreadFactory.java
+++ b/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/ClassLoaderThreadFactory.java
@@ -1,0 +1,99 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.clustering.jgroups;
+
+import java.security.PrivilegedAction;
+
+import org.jgroups.util.ThreadFactory;
+import org.wildfly.security.manager.WildFlySecurityManager;
+
+/**
+ * {@link ThreadFactory} decorator that associates a specific class loader to created threads.
+ * @author Paul Ferraro
+ */
+public class ClassLoaderThreadFactory implements org.jgroups.util.ThreadFactory {
+    private final ThreadFactory factory;
+    private final ClassLoader loader;
+
+    public ClassLoaderThreadFactory(ThreadFactory factory, ClassLoader loader) {
+        this.factory = factory;
+        this.loader = loader;
+    }
+
+    @Override
+    public Thread newThread(Runnable r) {
+        return this.newThread(r, null);
+    }
+
+    @Override
+    public Thread newThread(final Runnable r, String name) {
+        Runnable task = () -> {
+            try {
+                r.run();
+            } finally {
+                // Defensively reset the TCCL
+                this.setContextClassLoader(Thread.currentThread());
+            }
+        };
+        return this.setContextClassLoader(this.factory.newThread(task, name));
+    }
+
+    private Thread setContextClassLoader(Thread thread) {
+        PrivilegedAction<Thread> action = () -> {
+            thread.setContextClassLoader(this.loader);
+            return thread;
+        };
+        return WildFlySecurityManager.doUnchecked(action);
+    }
+
+    @Deprecated
+    @Override
+    public Thread newThread(ThreadGroup group, Runnable r, String name) {
+        return this.newThread(r, name);
+    }
+
+    @Override
+    public void setPattern(String pattern) {
+        this.factory.setPattern(pattern);
+    }
+
+    @Override
+    public void setIncludeClusterName(boolean includeClusterName) {
+        this.factory.setIncludeClusterName(includeClusterName);
+    }
+
+    @Override
+    public void setClusterName(String channelName) {
+        this.factory.setClusterName(channelName);
+    }
+
+    @Override
+    public void setAddress(String address) {
+        this.factory.setAddress(address);
+    }
+
+    @Override
+    public void renameThread(String base_name, Thread thread) {
+        this.factory.renameThread(base_name, thread);
+    }
+}

--- a/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/JChannelFactory.java
+++ b/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/JChannelFactory.java
@@ -92,10 +92,18 @@ public class JChannelFactory implements ChannelFactory, ProtocolStackConfigurato
         PrivilegedExceptionAction<JChannel> action = new PrivilegedExceptionAction<JChannel>() {
             @Override
             public JChannel run() throws Exception {
-                return new JChannel(JChannelFactory.this);
+                Thread thread = Thread.currentThread();
+                ClassLoader loader = thread.getContextClassLoader();
+                // We need to set the TCCL to be able to resolve custom ProtocolHook class names
+                thread.setContextClassLoader(JChannelFactory.class.getClassLoader());
+                try {
+                    return new JChannel(JChannelFactory.this);
+                } finally {
+                    thread.setContextClassLoader(loader);
+                }
             }
         };
-        final JChannel channel = WildFlySecurityManager.doChecked(action);
+        final JChannel channel = WildFlySecurityManager.doUnchecked(action);
         ProtocolStack stack = channel.getProtocolStack();
 
         TransportConfiguration transportConfig = this.configuration.getTransport();
@@ -231,6 +239,8 @@ public class JChannelFactory implements ChannelFactory, ProtocolStackConfigurato
             configureMulticastSocket(introspector, protocol, "diagnostics_addr", "diagnostics_port", diagnosticsSocketBinding);
         }
 
+        setProperty(introspector, protocol, "after_creation_hook", TransportConfigurator.class.getName());
+
         stack.add(protocol);
 
         final Class<? extends TP> transportClass = introspector.getProtocolClass().asSubclass(TP.class);
@@ -242,7 +252,7 @@ public class JChannelFactory implements ChannelFactory, ProtocolStackConfigurato
         };
 
         try {
-            stack.addAll(createProtocols(this.configuration, WildFlySecurityManager.doChecked(action).isMulticastCapable()));
+            stack.addAll(createProtocols(this.configuration, WildFlySecurityManager.doUnchecked(action).isMulticastCapable()));
         } catch (PrivilegedActionException e) {
             throw new IllegalStateException(e.getCause());
         }
@@ -368,7 +378,7 @@ public class JChannelFactory implements ChannelFactory, ProtocolStackConfigurato
                         return null;
                     }
                 };
-                WildFlySecurityManager.doChecked(action);
+                WildFlySecurityManager.doUnchecked(action);
             } catch (ClassNotFoundException e) {
                 throw new IllegalArgumentException(e);
             }

--- a/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/TransportConfigurator.java
+++ b/clustering/jgroups/extension/src/main/java/org/jboss/as/clustering/jgroups/TransportConfigurator.java
@@ -1,0 +1,49 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2016, Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags. See the copyright.txt file in the
+ * distribution for a full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.clustering.jgroups;
+
+import org.jgroups.protocols.TP;
+import org.jgroups.stack.Protocol;
+import org.jgroups.stack.ProtocolHook;
+import org.jgroups.util.DefaultThreadFactory;
+import org.jgroups.util.LazyThreadFactory;
+
+/**
+ * {@link ProtocolHook} that overrides the default thread factories for a transport.
+ * @author Paul Ferraro
+ */
+public class TransportConfigurator implements ProtocolHook {
+
+    @Override
+    public void afterCreation(Protocol protocol) throws Exception {
+        if (protocol instanceof TP) {
+            TP transport = (TP) protocol;
+            ClassLoader loader = JChannelFactory.class.getClassLoader();
+            transport.setThreadFactory(new ClassLoaderThreadFactory(new DefaultThreadFactory("", false), loader));
+            transport.setTimerThreadFactory(new ClassLoaderThreadFactory(new LazyThreadFactory("Timer", true, true), loader));
+            transport.setDefaultThreadPoolThreadFactory(new ClassLoaderThreadFactory(new DefaultThreadFactory("Incoming", false, true), loader));
+            transport.setOOBThreadPoolThreadFactory(new ClassLoaderThreadFactory(new DefaultThreadFactory("OOB", false, true), loader));
+            transport.setInternalThreadPoolThreadFactory(new ClassLoaderThreadFactory(new DefaultThreadFactory("INT", false, true), loader));
+        }
+    }
+}

